File: frontend/src/types/transaction.ts
/**
 * 交易 / 评分 / SHAP 类型
 * 对齐 D05 §4（交易反欺诈接口）与 baseline §4.1 transactions 表
 */
import type {
  Decision,
  RiskBand,
  TxType,
  Channel,
  ShapStatus,
  TaskStatus,
  FeedbackLabel
} from './enum'

/** 评分请求（D05 §4.1） */
export interface ScoreRequest {
  external_tx_id: string
  tx_type: TxType
  amount: number
  currency: string
  occurred_at: string
  card_token: string
  card_bin: string
  card_last4: string
  merchant_id?: string
  mcc?: string
  merchant_category?: string
  acquirer_id?: string
  device_fingerprint_hash?: string
  ip_address?: string
  ip_geo?: { country: string; city?: string }
  user_id: string
  user_created_at?: string
  channel?: Channel
  is_3ds_verified?: boolean
  shipping_country?: string
  billing_country?: string
  note_text?: string
  metadata?: Record<string, unknown>
}

/** 评分响应（D05 §4.1） */
export interface ScoreResponse {
  decision: Decision
  risk_score: number
  risk_band: RiskBand
  model_version: string
  rule_hits: RuleHit[]
  explainability: Explainability
  latency_ms: number
  case_id: string | null
  decision_id: string
}

export interface RuleHit {
  rule_id: string
  rule_name: string
  severity: 'INFO' | 'WARN' | 'BLOCK'
}

export interface Explainability {
  model_contribution: number
  rule_contribution: number
  shap_status: ShapStatus
  shap_task_id?: string
}

/** 异步评分任务（D05 §4.2-4.3） */
export interface AsyncScoreTask {
  task_id: string
  status: TaskStatus
  estimated_seconds?: number
  callback_event?: string
  result?: ScoreResponse
  created_at?: string
  completed_at?: string
}

/** 交易详情（D05 §4.6） */
export interface TransactionDetail {
  external_tx_id: string
  decision: Decision
  risk_score: number
  risk_band: RiskBand
  model_version: string
  rule_hits: RuleHit[]
  explainability: Explainability
  tx_type: TxType
  channel?: Channel
  is_3ds_verified?: boolean
  user_created_at?: string
  acquirer_id?: string
  shipping_country?: string
  billing_country?: string
  case_id?: string | null
  decision_id: string
  created_at: string
}

/** 交易列表查询参数（D05 §5.3） */
export interface TransactionQuery {
  external_tx_id?: string
  decision?: Decision
  risk_band?: RiskBand
  merchant_id?: string
  card_bin?: string
  model_version?: string
  start_time?: string
  end_time?: string
  min_amount?: number
  max_amount?: number
  channel?: Channel
  tx_type?: TxType
}

/** SHAP 计算结果（D05 §4.9） */
export interface ShapResult {
  shap_task_id: string
  decision_id: string
  model_id: string
  base_value: number
  prediction: number
  features: ShapFeature[]
  completed_at: string
}

export interface ShapFeature {
  name: string
  value: number
  shap: number
}

/** 反馈真实标签（D05 §4.5） */
export interface FeedbackRequest {
  external_tx_id: string
  label: FeedbackLabel
  label_source: 'CHARGEBACK' | 'MANUAL_REVIEW' | 'POLICE_REPORT' | 'EXTERNAL_LIST'
  labeled_at: string
  evidence?: string
}

/** 仪表盘 KPI 卡片数据（占位） */
export interface DashboardKpi {
  today_transactions: number
  blocked_count: number
  case_count: number
  model_auc: number
  p99_latency_ms: number
  drift_psi_7d: number
  fraud_loss_prevented_cents: number
  actual_loss_cents: number
  pass_rate: number
  appeal_count: number
}

File: backend/app/schemas/transaction.py
"""交易 schemas（D05 §4）。"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.common import Decision, RiskBand


class TxType(StrEnum):
    """交易类型（基准 §3.8）。"""

    PURCHASE = "PURCHASE"
    WITHDRAW = "WITHDRAW"
    REFUND = "REFUND"
    TRANSFER = "TRANSFER"
    TOPUP = "TOPUP"
    PAYMENT = "PAYMENT"


class Channel(StrEnum):
    """渠道（基准 §3.8）。"""

    WEB = "WEB"
    APP = "APP"
    POS = "POS"
    API = "API"
    QR = "QR"


class TransactionScoreRequest(BaseModel):
    """POST /transactions/score 请求体（D05 §4.1）。

    金额单位：分（int64），必须 > 0。
    card_token 为 Tokenization 后的 Token，不能传 PAN。
    """

    external_tx_id: str = Field(..., description="外部交易号，唯一")
    tx_type: TxType
    amount: int = Field(..., gt=0, description="金额（分）")
    currency: str = Field(default="CNY", min_length=3, max_length=3)
    occurred_at: datetime
    card_token: str
    card_bin: str = Field(..., min_length=6, max_length=6)
    card_last4: str = Field(..., min_length=4, max_length=4)
    merchant_id: str | None = None
    mcc: str | None = None
    merchant_category: str | None = None
    acquirer_id: str | None = None
    device_fingerprint_hash: str | None = None
    ip_address: str | None = None
    ip_geo: dict[str, Any] | None = None
    user_id: str
    user_account_id: str | None = None
    user_created_at: datetime | None = None
    channel: Channel | None = None
    is_3ds_verified: bool = False
    merchant_city: str | None = None
    shipping_country: str | None = Field(default=None, min_length=2, max_length=2)
    billing_country: str | None = Field(default=None, min_length=2, max_length=2)
    note_text: str | None = None
    metadata: dict[str, Any] | None = None
    # 异步深度分析
    analysis_depth: str | None = Field(default="STANDARD", description="STANDARD / DEEP")


class Explainability(BaseModel):
    """评分可解释性。"""

    model_contribution: float = 0.0
    rule_contribution: float = 0.0
    shap_status: str = "PENDING"
    shap_task_id: str | None = None


class RuleHit(BaseModel):
    """命中规则。"""

    rule_id: str
    rule_name: str | None = None
    severity: str | None = None


class TransactionScoreResponse(BaseModel):
    """POST /transactions/score 响应（D05 §4.1）。"""

    decision: Decision
    risk_score: float = Field(..., ge=0.0, le=1.0, description="0.0000-1.0000")
    risk_band: RiskBand
    model_version: str
    rule_hits: list[RuleHit] = Field(default_factory=list)
    explainability: Explainability = Field(default_factory=Explainability)
    latency_ms: int
    case_id: str | None = None
    decision_id: str


class TransactionDetail(BaseModel):
    """GET /transactions/{external_tx_id} 响应（D05 §4.6）。"""

    external_tx_id: str
    decision: Decision
    risk_score: float
    risk_band: RiskBand
    model_version: str
    rule_hits: list[RuleHit] = Field(default_factory=list)
    explainability: Explainability = Field(default_factory=Explainability)
    tx_type: TxType | None = None
    channel: Channel | None = None
    is_3ds_verified: bool = False
    user_created_at: datetime | None = None
    acquirer_id: str | None = None
    shipping_country: str | None = None
    billing_country: str | None = None
    case_id: str | None = None
    decision_id: str
    created_at: datetime


class AsyncScoreResponse(BaseModel):
    """POST /transactions/score/async 响应（D05 §4.2）。"""

    task_id: str
    status: str = "RUNNING"
    estimated_seconds: int = 30
    callback_event: str = "transaction.analysis_completed"


class BatchScoreRequest(BaseModel):
    """POST /transactions/score/batch 请求体（D05 §4.4）。"""

    transactions: list[TransactionScoreRequest] = Field(..., max_length=100)


class BatchScoreResultItem(BaseModel):
    """批量评分单项结果。"""

    external_tx_id: str
    decision: Decision
    risk_score: float
    risk_band: RiskBand
    error: str | None = None


class BatchScoreResponse(BaseModel):
    """批量评分响应。"""

    results: list[BatchScoreResultItem]
    success_count: int
    failure_count: int


class FeedbackRequest(BaseModel):
    """POST /transactions/feedback 请求体（D05 §4.5）。"""

    external_tx_id: str
    label: str = Field(..., description="FRAUD / NOT_FRAUD / SUSPECTED")
    label_source: str
    labeled_at: datetime
    evidence: str | None = None


__all__ = [
    "AsyncScoreResponse",
    "BatchScoreRequest",
    "BatchScoreResponse",
    "BatchScoreResultItem",
    "Channel",
    "Explainability",
    "FeedbackRequest",
    "RuleHit",
    "TransactionDetail",
    "TransactionScoreRequest",
    "TransactionScoreResponse",
    "TxType",
]
